### ISSUE TITLE
Allow search of history with regex argument

### ISSUE THREAD

**Issue by rocketraman** [fish-shell/fish-shell#3136]
It would be nice if when doing a history search, I could start from a regex. For example, if my history contained:

`foo bar baz`

I should be able to type:

`foo.*baz` with an `Alt-Up` [1] and find `foo bar baz`.

On a related note, even `history --search` doesn't seem to work with regexes -- one has to `grep` it for equivalent behavior.

[1] I'm not sure if this should be the default behavior or gated behind a separate key combo from a regular history traversal. Perhaps Ctrl-Alt-Up or something like that.

**Comment by krader1961** [comment]
I can see a strong argument for supporting regex searching via the `history` command. I'm ambivalent about complicating the interactive history searching with that capability. What happens when someone enters `foo*baz` thinking that globs rather than regexes are recognized? In fact, if we do implement this, I think the simpler glob syntax is preferable. As your example illustrates, what people normally want to do is limit the matches to sequences of strings separated by arbitrary characters. I would be very surprised if anyone would ever enter something like `abc[0-9]+ .*(def|ghi)` and press [up-arrow] to find history entries matching that pattern.

**Comment by rocketraman** [comment]
> As your example illustrates, what people normally want to do is limit the matches to sequences of strings separated by arbitrary characters.

Agreed. An alternate approach that makes this particular use case possible on the interactive history search is really what I was looking for, especially if the more general regex-based search is possibly on the `history` command. I generalized unnecessarily from my simple requirement to an interactive regex search.

**Comment by floam** [comment]
I dropped it off with the shipping guys at work. I'll check to make sure it
went out.

**Comment by sollidsnake** [comment]
I also like regexp searches, so I use this along with fish: https://github.com/junegunn/fzf
It brings to fish the ctrl-r hotkey with a nice interface.

**Comment by rocketraman** [comment]
> I also like regexp searches, so I use this along with fish: https://github.com/junegunn/fzf

Thanks for the pointer. Indeed, it seems quite useful. It has some issues integrating properly with fish, however. Example: https://github.com/junegunn/fzf/issues/436

Regardless, some basic interactive history improvements directly in fish would still be useful.

**Comment by krader1961** [comment]
FWIW, I don't believe we should ever implement glob or regex matching for auto-suggestions. That is, what you see when you type some characters and press [up-arrow], [ctrl-P], or otherwise invoke the `up-or-search` function. I do, on the other hand, believe we should support globs and PCREs for the `history search` command.

**Comment by krader1961** [comment]
In the past couple of months I've merged changes to the history subsystem to implement case-insensitive searching, display the timestamp, and support interactively deleting multiline commands. All of those should make it into the 2.4.0 release. So now I'm interested in resolving this issue.

What do people think about making `history search "term"` treat term as a glob by default rather than a literal string? We would also add `-r` and `--regex` flags to treat the search term as a PCRE pattern (same as `string match -r`)? Making glob matching the default is consistent with `string match` without `-r`. I don't think this will cause any problems in practice because it is pretty unlikely anyone is currently typing something like `history search 'abc*def'` expecting it to literally match that string.

**Comment by krader1961** [comment]
I've implemented the `history search` glob behavior in commit [e1315dc](https://github.com/krader1961/fish/commit/e1315dce9f0ab0200ba52c9f34074c150aff9203). Feel free to use or not as you see fit. TBD is `history search` PCRE behavior but I'm going to make that a separate issue in my fork since I think that is far less useful than simply allowing someone to type `history search 'abc*def'` to find every command containing those strings separated by some other strings.
